### PR TITLE
Google Cloud Platform (GCP) provider support.

### PR DESCRIPTION
**Pull request by ppknap**
This PR adds Google Cloud Platform support for cloud providers.
## Description

Following PR should be rather treated as _alpha_ version. It requires functional/integration tests which will be added after #9182 is done.

Blocked by #9184 
Related to #9134 #8165
## Motivation and Context

Possibility to use Google Cloud Platform in Koding.
## How Has This Been Tested?

Some basic unit tests were written. Additionally, templates and API calls were tested manually on Google platform.
## Types of changes
- [x] New feature (non-breaking change which adds functionality)
## Checklist:
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by rjeczalik — CHANGES_REQUESTED** [review]
```
go/src/vendor/golang.org/x/oauth2/google/default.go:17:2: cannot find package "cloud.google.com/go/compute/metadata" 
```

Dependant PR was merged, please rebase onto master.

**Comment by ppknap** [comment]
@rjeczalik thanks, will do! However, still waiting for the proper testing lib/env :)

File: go/src/koding/kites/kloud/provider/google/google.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	compute "google.golang.org/api/compute/v1"
+)
+
+var p = &provider.Provider{
+	Name:         "google",
+	ResourceName: "compute_instance",
+	Machine:      newMachine,
+	Stack:        nil,
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/google.go]: missing

File: go/src/koding/kites/kloud/provider/google/stack.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+				Network:      "default",
+				AccessConfig: struct{}{},
+			}
+		}
+
+		// means there will be several instances, we need to create a userdata
+		// with count interpolation, because each machine must have an unique
+		// kite id.
+		count := 1
+		if n, ok := instance["count"].(int); ok && n > 1 {
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/stack.go]: google provider does not support count attribute [0] like AWS one does, so there's no need to generate multiple keys for a single compute_instance resource (the ApplyTemplate will look more like vagrant than aws kloud provider) 

[0] https://www.terraform.io/docs/providers/google/r/compute_instance.html

File: go/src/koding/kites/kloud/provider/google/stack.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+		// root "metadata" object.
+		metadata := interfaceToMap(instance["metadata"])
+		s.Builder.InterpolateField(metadata, resourceName, "user-data")
+
+		// this part will be the same for all machines
+		userCfg := &userdata.CloudInitConfig{
+			Username: s.Req.Username,
+			Groups:   []string{"sudo"},
+			Hostname: s.Req.Username, // no typo here. hostname = username
+			KiteKey:  fmt.Sprintf("${lookup(var.%s, count.index)}", kiteKeyName),
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/stack.go]: after fixing https://github.com/koding/koding/pull/9185/files#r82092103 the kitekey can be set directly here

File: go/src/koding/kites/kloud/provider/google/stack.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+			for i := 0; i < count; i++ {
+				labels = append(labels, fmt.Sprintf("%s.%d", resourceName, i))
+			}
+		}
+
+		kiteKeyName := fmt.Sprintf("kitekeys_%s", resourceName)
+
+		// Cloud-Init can be injected only via "user-data" field defined in
+		// root "metadata" object.
+		metadata := interfaceToMap(instance["metadata"])
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/stack.go]: you can simplify this to:

```go
metadata, ok := instance["metadata"].(map[string]interface{})
if !ok {
    metadata = make(map[string]interface{})
}
```

File: go/src/koding/kites/kloud/provider/google/bootstrap.json.tmpl [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{
+  "provider": {
+    "google": {
+      "credentials": "${var.google_credentials}",
+      "project": "${var.google_project}",
+      "region": "${var.google_region}"
+    }
+  },
+  "output": {
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/bootstrap.json.tmpl]: The bootstrap template is expected to produce output variables - we need to output some variables here (the network id, firewall rule etc.) in order to set Bootstrap struct after bootstrapping.

The frontend calls bootstrap, when bootstrap struct is empty; if bootstrap struct is not updated after successful  bootstrap, then next time template is saved - the bootstrap is going to be called again.

File: go/src/koding/kites/kloud/provider/google/google.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+func (c *Cred) ComputeService() (*compute.Service, error) {
+	cfg, err := google.JWTConfigFromJSON([]byte(c.Credentials), compute.ComputeScope)
+	if err != nil {
+		return nil, err
+	}
+
+	return compute.New(cfg.Client(context.Background()))
+}
+
+type Bootstrap struct{}
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/google.go]: This should define at least one field that is going to be updated after bootstrap. Empty or nil Bootstrap means no bootstrap, and we do bootstrap for google provider.

File: go/src/koding/kites/kloud/provider/google/machine.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+		Start:        45,
+		Finish:       60,
+	}
+
+	return nil, ws.Wait()
+}
+
+// Start stops Google compute instance.
+func (m *Machine) Stop(ctx context.Context) (interface{}, error) {
+	ev, withPush := eventer.FromContext(ctx)
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/machine.go]: We could move this to BaseMachine to reduce further the provider boilerplate. Commenting for posterity.

File: go/src/koding/kites/kloud/provider/google/stack.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/kites/kloud/provider/google/stack.go]: it does support `count` attribute. `count` is a terraform [property](https://www.terraform.io/docs/configuration/resources.html#count) applicable for most terraform resources. 

Here is an example for GCP:

```sh
(archpk) ~/terraform/gce_tests$ grep -C 3 "count" gce_stack.tf.json 
  "resource": {
    "google_compute_instance": {
      "cluster": {
        "count": 2,
        "name": "gce-${var.environment_name}-instance",
        "machine_type": "f1-micro",
        "zone": "us-central1-a",
(archpk) ~/terraform/gce_tests$ terraform plan | grep -A 2 "google_compute_instance"
google_compute_instance.cluster.0: Refreshing state... (ID: gce-development-instance)

The Terraform execution plan has been generated and is shown below.
--
+ google_compute_instance.cluster.0
    can_ip_forward:                                      "false"
    disk.#:                                              "1"
--
+ google_compute_instance.cluster.1
    can_ip_forward:                                      "false"
    disk.#:                                              "1"
```

Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/stack.go]: TIL thanks for clarification

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/kites/kloud/provider/google/google.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		}
+	}
+
+	return fmt.Errorf("unknown region name: %v", r)
+}
+
+// Cred represents jCredentialDatas.meta for "google" provider.
+type Cred struct {
+	Credentials string `json:"credentials" bson:"credentials" hcl:"credentials" kloud:",secret"`
+	Project     string `json:"project" bson:"project" hcl:"project" kloud:",secret"`
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/provider/google/google.go]: This shouldn't be ,secret.

**Comment by rjeczalik** [comment]
👍

**Pull request merged**